tests/ratingservice/main_test.py

```diff
@@ -44,25 +45,36 @@ def setUp(self):
     def tearDown(self):
         self.session.close()
 
-    def testGetRating(self):
-        """ test getting ratings for all shop products """
+    def testGetAllRatings(self):
+        """ test getting all ratings """
+        url = composeUrl("ratings")
+        res = self.session.get(url, timeout=TIMEOUT)
+        self.assertEqual(res.status_code, 200)
+        ratings = res.json().get('ratings')
+        self.assertTrue(type(ratings) == list)
+        self.assertEqual(len(ratings), len(products))
+        ids = list(map(lambda r: r['id'], ratings))
+        self.assertTrue(all(p in products for p in ids))
+
+    def testGetRatingPerProduct(self):
+        """ test getting ratings for each shop product """
         for product in products:
             url = composeUrl("rating", product)
-            res = self.session.get(url, timeout=1.0)
+            res = self.session.get(url, timeout=TIMEOUT)
```

Comment:
I'm not sure if I like how the products list is dynamically loaded. If there's an error and products isn't loaded properly, some of these tests will mistakenly pass. A couple thoughts:
- can we hardcode the product list with the expected data?
- If not, `products` should probably at least be loaded in the setup() function instead of `__main__`
- There should probably at least be some assertions on `products` after it is loaded, to make sure it has reasonable data

Reply:
i do not follow. are you referencing the fact that we already test the same in the previous test method? the matter is that another method was just added. before that there was no way to test that all shop products have rating.
now, we can reduce the test for GET /product/<id> for testing a custom product from the list.
answering your questions:
* product list is hardcoded but not the ratings. the ratings are generated at each deployment. we test rating correctness in the last test which includes both the calculation and the correctness.
* list of products is loaded from the products.json at the beginning of the test. if you can explain the difference between the exact place where the code for loading the products is located, it will help me to understand your arguments.
* this last bullet is unclear. the products are loaded from the same source from which the shop loads it. there should be no need for additional assertion i think.

Reply:
> list of products is loaded from the products.json at the beginning of the test

Ah ok, when I skimmed the code I thought it was querying it from the service. You can ignore my first point then

> list of products is loaded from the products.json at the beginning of the test.

For my second point, if I'm reading it right, products is populated by `getProducts()` in part of `if __name__ == '__main__'`. This means it will be populated if you run the script directly, but if you call it using pytest (ex, by running `pytest ./`), that `__main__` block will be skipped, right?

> this last bullet is unclear. the products are loaded from the same source from which the shop loads it. there should be no need for additional assertion i think.

For my third bullet, I was pointing out that your tests take as a given that products contains items, but it is initialized as an empty array. Some tests will silently pass if products is never populated, and others will fail in unexpected places. It's probably worthwhile to add an assertion at the start of each test to make sure products contains some data

Reply:
ok, thanx. regarding the pytest, it will not work because it expects to receive arguments and will fail given the arguments aren't provided, the arguments are the root URL of the service and the path to products.json. also the second can be defaulted. what is a way to pass the arguments using pytest? is it a blocker for this PR?
for the last bullet, the comparison goes the returned list vs the list from the product.json. The second list cannot be empty, hence the test will fail if the returned list is empty.

Reply:
can i mark it resolved given i will take care to support pytest in one of further PRs?

Reply:
> is the use pytest is a hard requirement at the moment? i see that most of the tests use unittest and run by python.
it might be better to define a migration to pytest as a separate task and refactor all the modules at once. wdyt?

Is there a reason you can't just put `getProducts` in a set up function instead of as part of the `__main__` function? That's the standard practice for python tests and it seems like a simple enough change. Are there downsides I'm not thinking of?

> for the last bullet, the test compares the returned list vs. the list read from the product.json file. The second list cannot be empty, hence the test will fail if the returned list is empty.

You're assuming the second list can't be empty ;-) I think it makes sense to assert those assumptions as part of the test. If the `__main__` function isn't executed, or the the product.json file is replaced with an empty file, `products` will be empty, so I think it's reasonable to add an assertion to the top of the test to check for that. Is there a reason you'd rather not?

Reply:
this is a very long thread. let's have a quick sync over GVC to talk about these items.

Reply:
I've looked into python unittest options and used `setUpClass` to initiate the values and fail() to handle possible problems when reading products and/or passing service URL.
No code except for
```python
if __name__ == '__main__':
    unittest.main(argv=['first-arg-is-ignored'])
```
is not defined outside the class.